File: crates/api/src/http/dto/response/signal.rs
//! Response DTO for the signal feed.

use chrono::{DateTime, Utc};
use serde::Serialize;

use crate::application::EnrichedSignal;
use crate::http::dto::EmbeddedTokenResponse;

/// `GET /api/signals` item (and `GET /api/signals/stream` event — both
/// endpoints emit exactly this object).
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SignalResponse {
    /// Storage identity — stable across pages, usable as a client-side
    /// list key. Emitted as a JSON number: signals are sparse, the
    /// sequence will never approach 2^53.
    pub(crate) id: i64,
    pub(crate) detector: String,
    pub(crate) protocol: String,
    pub(crate) pool_address: String,
    /// The pool's token pair, embedded in the same shape as
    /// `PoolResponse` (`EmbeddedTokenResponse`). Every field inside is
    /// nullable — a pool not yet resolved by yog-context embeds the
    /// minimal view, and the client falls back to `pool_address`.
    pub(crate) token_a: EmbeddedTokenResponse,
    pub(crate) token_b: EmbeddedTokenResponse,
    pub(crate) severity: String,
    // Exact decimals are emitted as strings, like every other decimal
    // quantity of the API (a JSON-number consumer would go through
    // lossy f64).
    pub(crate) value: String,
    pub(crate) threshold: Option<String>,
    pub(crate) message: Option<String>,
    pub(crate) triggered_at: DateTime<Utc>,
}

impl From<EnrichedSignal> for SignalResponse {
    fn from(enriched: EnrichedSignal) -> Self {
        let signal = enriched.record.signal;
        Self {
            id: enriched.record.id,
            detector: signal.detector,
            protocol: signal.protocol.to_string(),
            pool_address: signal.pool_address.to_string(),
            token_a: enriched.token_a.into(),
            token_b: enriched.token_b.into(),
            severity: signal.severity.to_string(),
            value: signal.value.to_string(),
            threshold: signal.threshold.map(|t| t.to_string()),
            message: signal.message,
            triggered_at: signal.triggered_at,
        }
    }
}

File: web/src/lib/api/schema/signal.ts
import * as z from "zod";
import { BigDecimal, Rfc3339, SignedBigDecimal } from "./shared";
import { TokenSchema } from "./token";

// ─────────────────────────────────────────────────────────────────────
// SignalResponse — mirrors `api::http::dto::response::SignalResponse`
// ─────────────────────────────────────────────────────────────────────

/**
 * Severity levels, mirroring the Rust `Severity` enum (and the CHECK
 * constraint on the `signals` table). A closed set — an unknown value
 * is a wire-contract drift worth failing on.
 */
export const SeveritySchema = z.enum(["info", "warning", "critical"]);
export type Severity = z.infer<typeof SeveritySchema>;

/**
 * Wire shape of a single signal, shared by the paginated feed
 * (`GET /api/signals`) and the SSE stream (`GET /api/signals/stream`)
 * — both endpoints emit exactly this object.
 */
export const SignalSchema = z.object({
  // Storage identity: stable across pages and stream events, used as
  // the reconciliation key when the live feed refetches after a
  // reconnect (and as the React list key).
  id: z.number().int(),
  detector: z.string().min(1),
  protocol: z.string().min(1),
  poolAddress: z.string().min(1),
  // The pool's token pair, embedded in the same shape as PoolResponse
  // (`EmbeddedTokenResponse` on the wire). Every field inside is
  // nullable — an unresolved pool embeds the minimal view and the UI
  // falls back to the pool address.
  tokenA: TokenSchema,
  tokenB: TokenSchema,
  severity: SeveritySchema,
  // The metric that crossed the threshold. Signed: a deviation can be
  // below as well as above the reference.
  value: SignedBigDecimal,
  // The configured threshold it crossed (always positive), when the
  // detector has a single scalar one.
  threshold: BigDecimal.nullable(),
  message: z.string().nullable(),
  triggeredAt: Rfc3339,
});

export type SignalResponse = z.infer<typeof SignalSchema>;
